boa_engine/src/builtins/uri/mod.rs

```diff
@@ -0,0 +1,553 @@
+                if n == 1 || n > 4 {
+                    context.throw_uri_error("TO-DO")?;
+                }
+
+                // 2. If k + (3 × (n - 1)) ≥ strLen, throw a URIError exception.
+                if k + (3 * (n - 1)) > str_len {
+                    context.throw_uri_error("TO-DO")?;
+                }
+
+                // 3. Let Octets be « B ».
+                let mut octets = vec![b];
+
+                // 4. Let j be 1.
+                // 5. Repeat, while j < n,
+                for _j in 1..n {
+                    // a. Set k to k + 1.
+                    k += 1;
+
+                    // b. If the code unit at index k within string is not the code unit 0x0025 (PERCENT SIGN), throw a URIError exception.
+                    if code_units[k] != 0x0025 {
+                        context
+                            .throw_uri_error("escape characters must be preceded with a % sign")?;
+                    }
+
+                    // c. If the code units at index (k + 1) and (k + 2) within string do not represent hexadecimal digits, throw a URIError exception.
+                    let unit_k_1 = code_units[k + 1];
+                    let unit_k_2 = code_units[k + 2];
+                    if !is_hexdigit(unit_k_1) || !is_hexdigit(unit_k_2) {
+                        context.throw_uri_error("invalid escape character")?;
+                    }
+
+                    // d. Let B be the 8-bit value represented by the two hexadecimal digits at index (k + 1) and (k + 2).
+                    let b = decode_byte(unit_k_1, unit_k_2);
+
+                    // e. Set k to k + 2.
+                    k += 2;
+
+                    // f. Append B to Octets.
+                    octets.push(b);
+
+                    // g. Set j to j + 1.
+                }
+
+                // 6. Assert: The length of Octets is n.
+                assert_eq!(octets.len(), n);
+
+                // 7. If Octets does not contain a valid UTF-8 encoding of a Unicode code point, throw a URIError exception.
+                match String::from_utf8(octets) {
+                    Err(_) => {
+                        return Err(context.construct_uri_error("invalid UTF-8 encoding found"))
+                    }
+                    Ok(v) => {
+                        // 8. Let V be the code point obtained by applying the UTF-8 transformation to Octets, that is, from a List of octets into a 21-bit value.
+
+                        // 9. Let S be UTF16EncodeCodePoint(V).
+                        // utf16_encode_codepoint(v)
+                        v.encode_utf16().collect::<Vec<_>>()
+                    }
+                }
+            }
+        };
+
+        // e. Set R to the string-concatenation of R and S.
+        r.extend_from_slice(&s);
+
+        // f. Set k to k + 1.
+        k += 1;
+    }
+}
+
+/// Checks if a given code unit is an hexadecimal digit represented in UTF-16.
+fn is_hexdigit(code_unit: u16) -> bool {
```

Comment:
Instead of returning a bool here, this could return `Option<u8>` since we check if it is hex (in u8 range). Then the resulting `u8`s could be passed to `decode_byte` and that function would be safer, because it would not contain the `u8::try_from`.

Or even better, compine `is_hexdigit` and `decode_byte`. They are only called directly after each other and if we combine them we can avoid the double check and all potential panics.

Reply:
I merged both functions, let me know what you think :)

Reply:
I was thinking of doing it like this:

```rust
fn decode_hex_byte(high: u16, low: u16, context: &mut Context) -> JsResult<u8> {
    match (u8::try_from(high), u8::try_from(low)) {
        (Ok(high), Ok(low)) => {
            let high = if (b'0'..=b'9').contains(&high) {
                high - b'0'
            } else if (b'A'..=b'F').contains(&high) {
                high - b'A' + 0x0A
            } else if (b'a'..=b'f').contains(&high) {
                high - b'a' + 0x0a
            } else {
                return Err(context.construct_uri_error("invalid hexadecimal digit found"));
            };

            let low = if (b'0'..=b'9').contains(&low) {
                low - b'0'
            } else if (b'A'..=b'F').contains(&low) {
                low - b'A' + 0x0A
            } else if (b'a'..=b'f').contains(&low) {
                low - b'a' + 0x0a
            } else {
                return Err(context.construct_uri_error("invalid hexadecimal digit found"));
            };

            Ok((high << 4) + low)
        }
        _ => Err(context.construct_uri_error("invalid hexadecimal digit found")),
    }
}
```

That should avoid any double check and there is no panic anymore.

Reply:
Done :)